src/cloud-api-adaptor/pkg/userdata/provision.go

```diff
@@ -196,13 +230,78 @@ func processCloudConfig(cfg *Config, cc *CloudConfig) error {
 		{path: cfg.paths.authJson, optional: true},
 	}
 
+	// entries pre-defined
 	for _, e := range entries {
-		err := e.writeFile(cc)
+		bytes := findConfigEntry(e.path, cc)
+		if bytes == nil {
+			if !e.optional {
+				return fmt.Errorf("failed to find %s entry in cloud config", e.path)
+			}
+			continue
+		}
+		err := writeFile(e.path, bytes)
 		if err != nil {
 			return err
 		}
 	}
 
+	// entries not pre-defined by caa, we have some special config for some specific providers, handle it here...
+	for _, wf := range cc.WriteFiles {
+		path := wf.Path
+		bytes := []byte(wf.Content)
+		if bytes != nil && !isPredefinedConfigEntry(path, entries) {
+			if err := writeFile(path, bytes); err != nil {
+				return fmt.Errorf("failed to write config file %s: %w", path, err)
+			}
+		}
+	}
+
+	return nil
+}
+
+func calculateUserDataHash(cfg *Config) error {
+	initToml, err := os.ReadFile(cfg.initdataMeta)
+	if err != nil {
+		return err
+	}
+	var initdata InitData
+	err = toml.Unmarshal(initToml, &initdata)
+	if err != nil {
+		return err
+	}
+
+	checksumStr := ""
+	var byteData []byte
+	for _, file := range cfg.staticFiles {
+		if _, err := os.Stat(file); err == nil {
+			logger.Printf("calculateUserDataHash and reading file %s\n", file)
+			bytes, err := os.ReadFile(file)
+			if err != nil {
+				return fmt.Errorf("Error reading file %s: %v", file, err)
+			}
+			byteData = append(byteData, bytes...)
```

Comment:
can we use a hash extension algorithm (like in a TEE launch measurements or TPM registers) if we just concat all bytes the digest of file1 "abc" + file2 "def" will be the same as the digest of file1 "ab" and file2 "cdef".

it could also define the order in which the digest is calculated, if we use a hash extension algorithm:

```
zeros = [0;32];
previous_digest = zeros
for file in files:
  previous_digest = sha256_digest(previous_digest + bytes(file))
```

Reply:
I agree `re-construct the original initdata toml from the annotation 1:1 in process-user-data and hash it` might be a better approach.

Reply:
yes, if we reconstruct the full initdata toml, there is no need for hash extension, because there is no ambiguity.

Reply:
@mkulke I'm keeping existing algorithm for 2 reasons: 1. The toml marshall can not guarantee the 1:1 format when reconstruct the initdata as you also found. 2. previous-digest+contents makes it difficult to calculate the hash value manually. So I updated the md file to guide users how to calculate the hash value manually in https://github.com/confidential-containers/cloud-api-adaptor/pull/1912/commits/687fb90160c467008271fc85a442d9532cce8d14

Reply:
I am not sure it's a good idea to have a special hash algorithm for Peerpods. This will be confusing to users (and future initdata tooling). As far as I can see, we don't gain anything from splitting the initdata.toml into individual files on the node, we could just split the initdata on the guest (in process-user-data) and calculate the hash of the full body.

Reply:
@mkulke do you mean create a file like
```yaml
write_files:
- path: /run/peerpod/daemon.json
  content: 
- path: /run/peerpod/initdata.toml
  content:
 ```
And then use `process-user-data` to parse `/run/peerpod/initdata.toml`? I guess it's doable for providers which still depend on cloud-init.

Reply:
@mkulke I dropped a commit to use the annotation for initdata directly here https://github.com/confidential-containers/cloud-api-adaptor/pull/1912/commits/96d3466809c93711b07f7eac19837a581af074c7, wdyt this approach?